### PR TITLE
[IOTDB-2307]UDF Library Series Discovery Functions

### PR DESCRIPTION
**Pull request by trin1t**
## Description
UDF Library sereis discovery functions, tests and documents

### Function list
ConsecutiveSequences
ConsecutiveWindows

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: library-udf/src/main/java/org/apache/iotdb/library/util/Util.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        ans = row.getDouble(0);
+        break;
+      case BOOLEAN:
+        ans = row.getBoolean(0);
+        break;
+      case TEXT:
+        ans = row.getString(0);
+        break;
+    }
+    return ans;
```
Comment (wangchao316) [inline library-udf/src/main/java/org/apache/iotdb/library/util/Util.java:103]: Whether to return -1 if the case does not meet the requirement

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+    String sqlStr = "select ConsecutiveSequences(d1.s1,d1.s2) from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
+        Statement statement = connection.createStatement()) {
+      ResultSet resultSet = statement.executeQuery(sqlStr);
+      int resultSetLength = resultSet.getRow();
+      assert resultSetLength == 3;
```
Comment (wangchao316) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:200]: please use assertEquals

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        .setUdfCollectorMemoryBudgetInMB(5)
+        .setUdfTransformerMemoryBudgetInMB(5)
+        .setUdfReaderMemoryBudgetInMB(5);
+    EnvFactory.getEnv().initBeforeClass();
+    createTimeSeries();
+    generateData();
+    registerUDF();
+  }
+
+  private static void createTimeSeries() throws MetadataException {
```
Comment (wangchao316) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:65]: I found that there are several classes that create sequences and prepare data, and can I extract them to one place for common use.

**Review by wangchao316 — CHANGES_REQUESTED** [review]
Good Job

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 2 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been changed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 3 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:74]: The structure of data generation is modified from integration tests, where every test prepares data seperately. If it is necessary, I can update it in the next version.

File: library-udf/src/main/java/org/apache/iotdb/library/util/Util.java (hunk 1 above)
Comment (trin1t) [reply library-udf/src/main/java/org/apache/iotdb/library/util/Util.java:103]: It already covers all data types. Perhaps throw an error when none is satisfied?

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        TSDataType.DOUBLE,
+        TSEncoding.PLAIN,
+        CompressionType.UNCOMPRESSED,
+        null);
+  }
+
+  private static void generateData() {
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:105]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveSequences1() {
+    String sqlStr = "select ConsecutiveSequences(d1.s1,d1.s2) from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:205]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveSequences2() {
+    String sqlStr = "select ConsecutiveSequences(d2.s1,d2.s2) from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:241]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveSequences3() {
+    String sqlStr = "select ConsecutiveSequences(d1.s1,d1.s2,'gap'='5m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:277]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveSequences4() {
+    String sqlStr = "select ConsecutiveSequences(d2.s1,d2.s2,'gap'='5m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:313]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveWindows1() {
+    String sqlStr = "select ConsecutiveWindows(d1.s1,d1.s2,'length'='10m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:349]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+  }
+
+  @Test
+  public void testConsecutiveWindows2() {
+    String sqlStr = "select ConsecutiveWindows(d2.s1,d2.s2,'length'='10m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:385]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+
+  @Test
+  public void testConsecutiveWindows3() {
+    String sqlStr =
+        "select ConsecutiveWindows(d1.s1,d1.s2,'length'='10m','gap'='5m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:422]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+
+  @Test
+  public void testConsecutiveWindows4() {
+    String sqlStr =
+        "select ConsecutiveWindows(d2.s1,d2.s2,'length'='10m','gap'='5m') from root.vehicle";
+    long timeStamp = 0;
+    int value = 0;
+    try (Connection connection =
+            DriverManager.getConnection(
+                Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
```
Comment (ericpai) [inline library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java:459]: Please use `Connection connection = EnvFactory.getEnv().getConnection();` instead.

**Review by ericpai — CHANGES_REQUESTED** [review]

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 4 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 5 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 6 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 7 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 8 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 9 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 10 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 11 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java (hunk 12 above)
Comment (trin1t) [reply library-udf/src/test/java/org/apache/iotdb/library/series/SeriesTest.java]: All have been fixed.

File: library-udf/src/main/java/org/apache/iotdb/library/util/LinearRegression.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+ * regular usage.
+ */
+public class LinearRegression {
+  double[] x, y, e, yhead;
+  int n;
+  double sumx, sumy, xbar, ybar, xxbar, yybar, xybar;
+  double beta1, beta0, rss, ssr, R2, svar, svar1, svar0;
+
+  public LinearRegression(double[] a, double[] b) {
+    x = a.clone();
```
Comment (ericpai) [inline library-udf/src/main/java/org/apache/iotdb/library/util/LinearRegression.java:34]: Does it need to check n should be larger than 0?

Comment (trin1t) [reply library-udf/src/main/java/org/apache/iotdb/library/util/LinearRegression.java:34]: Checks are added in data profiling part. There is no usage in series discovery functions.

**Review by ericpai — APPROVED** [review]

**Review by wangchao316 — APPROVED** [review]
LGTM

**Pull request merged**